rclc_parameter/src/rclc_parameter/parameter_server.c

```diff
@@ -0,0 +1,618 @@
+      rclc_parameter_service_publish_event(parameter_server);
+    }
+
+    if (parameter_server->on_modification) {
+      parameter_server->on_modification(parameter);
+    }
+
+    return RCL_RET_OK;
+  }
+}
+
+rcl_ret_t rclc_parameter_set_double(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name,
+  double value)
+{
+  Parameter * parameter =
+    rclc_parameter_search(&parameter_server->parameter_list, parameter_name);
+
+  if (parameter == NULL) {
+    return RCL_RET_ERROR;
+  }
+
+  if (parameter->value.type != RCLC_PARAMETER_DOUBLE) {
+    return RCL_RET_INVALID_ARGUMENT;
+  } else {
+    parameter->value.double_value = value;
+
+    if (parameter_server->notify_changed_over_dds) {
+      rclc_parameter_prepare_parameter_event(&parameter_server->event_list, parameter, false);
+      rclc_parameter_service_publish_event(parameter_server);
+    }
+
+    if (parameter_server->on_modification) {
+      parameter_server->on_modification(parameter);
+    }
+
+    return RCL_RET_OK;
+  }
+}
+
+rcl_ret_t rclc_parameter_get_bool(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name,
+  bool * output)
+{
+  Parameter * parameter =
+    rclc_parameter_search(&parameter_server->parameter_list, parameter_name);
+
+  rcl_ret_t ret = RCL_RET_OK;
+
+  if (parameter == NULL) {
+    ret = RCL_RET_ERROR;
+  } else if (parameter->value.type != RCLC_PARAMETER_BOOL) {
+    ret = RCL_RET_INVALID_ARGUMENT;
+  } else if (parameter != NULL) {
+    *output = parameter->value.bool_value;
+  }
+
+  return ret;
+}
+
+rcl_ret_t rclc_parameter_get_int(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name,
+  int * output)
+{
+  Parameter * parameter =
+    rclc_parameter_search(&parameter_server->parameter_list, parameter_name);
+
+  rcl_ret_t ret = RCL_RET_OK;
+
+  if (parameter == NULL) {
+    ret = RCL_RET_ERROR;
+  } else if (parameter->value.type != RCLC_PARAMETER_INT) {
+    ret = RCL_RET_INVALID_ARGUMENT;
+  } else if (parameter != NULL) {
+    *output = parameter->value.integer_value;
+  }
+
+  return ret;
+}
+
+rcl_ret_t rclc_parameter_get_double(
+  rclc_parameter_server_t * parameter_server,
+  const char * parameter_name,
+  double * output)
+{
+  Parameter * parameter =
+    rclc_parameter_search(&parameter_server->parameter_list, parameter_name);
+
+  rcl_ret_t ret = RCL_RET_OK;
+
+  if (parameter == NULL) {
+    ret = RCL_RET_ERROR;
+  } else if (parameter->value.type != RCLC_PARAMETER_DOUBLE) {
+    ret = RCL_RET_INVALID_ARGUMENT;
+  } else if (parameter != NULL) {
+    *output = parameter->value.double_value;
+  }
+
+  return ret;
+}
+
+rcl_ret_t rclc_parameter_service_publish_event(
+  rclc_parameter_server_t * parameter_server)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(&parameter_server->event_publisher, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(&parameter_server->event_list, RCL_RET_INVALID_ARGUMENT);
+
+  struct timespec ts;
+  clock_gettime(CLOCK_REALTIME, &ts);
```

Comment:
Use `rcutils_system_time_now` from https://github.com/micro-ROS/rcutils/blob/foxy/include/rcutils/time.h for compatibility?

Reply:
8e465f8204365d555eb277439abf7f39dd80d198